### ISSUE TITLE
Flagging geomagnetic data and how to include that into data formats

### ISSUE THREAD

**Issue by leonro** [INTERMAGNET/wg-www-gins-data-formats#3]
## Add flagging information to geomagnetic data sets 

Background: do not remove data, but mark valid or invalid data so that the decision gets transparent:  
-	Flagging will lead to reproducible data selection criteria.
-	Flagging will allow for selecting specifically flagged data for in-depth  research (e.g. pulsations, lightning)
-	Flagging information can be stored and distributed along with data sets (requiring ImagCDF format).
This document will provide an overview about current discussions and developments on how to integrate such information into a normal workflow.

## 1. Introduction

INTERMAGNET has decided that optional flagging information is a valuable extension to be provided along with geomagnetic data sets. Flagged data is not “cleaned” any more in a traditional sense by removing disturbed data, but data of interest is marked valid or invalid with flags. Therefore observer decisions get transparent. In this document we will briefly discuss how such flagging information can be added to geomagnetic data sets, way this should be done and how this information can be stored and reconstructed. Although optional such procedure has a number of benefits:  
-	Flagging will lead to reproducible data selection criteria.
-	Flagging will allow for selecting specifically flagged data for in-depth research (e.g. pulsations, lightning), which partly would have been removed from the data sets.
-	Flagging information can be stored and distributed along with data sets (requiring ImagCDF format).

### 2.1 Flags at the Conrad Observatory
The flagging procedure at the Conrad Observatory allows the observer to mark specific data points or ranges. Flags are used for labelling data spikes, storm onsets, pulsations, disturbances, lightning strikes, etc. Each flag is associated with a comment and a type number. The flagtype number ranges between 0 and 4:

    0: normal data with comment (e.g. " Workshop at the Observatory")
    1: data marked by automated outlier analysis (e.g. spikes)
    2: data marked by an observer as valid geomagnetic signature (e.g. storm onset, pc4 pulsation). Such data cannot be marked invalid by automated procedures
    3: data marked by an observer as invalid for definite data (e.g. lightning, magnetic disturbance)
    4: merged data (e.g. data inserted from another source/instrument as defined in the comment)

Flags can be stored along with the data set (requires CDF format output) or separately in a binary archive. These flags can then be applied to the raw data again, ascertaining perfect reproducibility.

### 2.2 Flags at the IPGP 
The flagging procedure at the Chambon la forêt Observatory will be applied on all IPGP network. It will allow the observer to mark specific data points or ranges. Flags will be used for labelling data spikes, storm onsets, pulsations, disturbances, lightning strikes, etc. For now, our early discussions contemplate the flagtype number ranges between 0 and 9:

    0: data unchanged
    1: shifted data
    2: interpolated data
    3: merged data (e.g. data inserted from another source/instrument)
    4: noteworthy data (pulsation, SSC, storms etc) 
    5 to 9: data marked by an observer as invalid for definite data (e.g. lightning, magnetic disturbance, abnormal value generated by equipment, human causes like vehicles, interference signal and “unexplained”). This data is not used for various mean calculations (minute, hourly, daily, monthly and annual means)

 
It is hoped flags can be stored along with the data set (requires CDF format output).

A typical flag can consist of the following information of which elements marked with * are considered optional.   
    a) Begin:
    b) End:
    c) Affected Component(s):
    d) Code:
    e) Comment:
    f) *ModificationDate:
    g) *User:
    h) *Affected System/Sensor: 

### 2.6 Suggestions for an INTERMAGNET Flagging Ruleset
One could provide a thesauris type of dictionary, linking id numbers to certain events: e.g.

    0001 : lightning strike
    0354 : 

## 3. Applying Flags to data
Flagging can be performed manually or supported by software routines. When using MagPy refer to the manual (https://github.com/geomagpy/magpy) section 2.5.

## 4. Writing flags to ImagCDF

-	Maintain optional state of flagging procedure 
-	Data can be stored and read either with or without flagging information
-	Flagging-Ruleset-Variable allows to use and define specfic rulesets. This technique allows to intergrate already existing flags (IPGP, USGS, Niemegk, Conrad) and to develop a INTERMAGNET specific version. Only necessary is an interpreter. The suggested content information (section 2) is sufficient to comprise all of the already existing rulesets. 
-	Flagging information can be added and modified any time to already existing data sets.
-	Memory efficient storing (as preferred for Archive files) can be accomplished by adding flag information as an array.
-	Flags can be extracted without reading the full data set -> efficient search for specific information

Field to be considered for ImagCDF attributes:

| Attribute Name	| Type	| Entries	| Description |
| ------------- | ------------- | ------------- | ------------- |
| FlagRulesetType	| String	| 1	| A string defining the ruleset type of the flagging information (e.g. Conrad, IMFlags etc)| 
| FlagRulesetVersion	| String	| 1	| A version code for each ruleset e.g. 1.0| 

Flagging information is stored in within the following data fields:

| Field Name	| Type	| Required 	| Description| 
| ------------- | ------------- | ------------- | ------------- |
| FlagBeginTimes	| TT2000/Datetime	| All	| A time object defining the onset of the flag | 
| FlagEndTimes	| TT2000/Datetime	| 	| A time object defining the end time of the flag | 
| FlagComponents	| String	| 	| Reference to the field component to which the flag should be added (e.g. GeomagneticFieldX) | 
| FlagCode	| Int	| 	| A identification number describing the type of flag| 
| FlagComments	| String	| 	| A free text information on the flag| 
| FlagSystemReference	| String	| 	| Reference to the system/datafile, to which the flag should be applied to. Helpful, if only flagging information is extracted.| 
| FlagObserver	| String	| 	| Name of the Observer, who added/modified the flag| 
| FlagModificationTimes	| TT2000/Datetime	| 	| A version code for each ruleset e.g. 1.0| 

“Required” denotes the rulesets for which this column is required for. An asterix indicates optional inputs.

**Comment by CharlesBlais** [comment]
Hi all, regarding flagging geomagnetic data and including it into data formats, is there any comments regarding @leonro suggestions?  Regarding the proposed ruleset, defining a rule set in numeric representation as you provided is great.  String representation wouldn't be possible in certain storage data formats of certain institutes.  Speaking for Canada, we don't keep these level of details so it would be a challenge to incorporate all this (but I would leave it to David Calp).  Is this a topic we should be opening up beyond just GWD members?

**Comment by sputnik-a** [comment]
Thanks @leonro and @vmaury-ipgp for starting the discussion and preparing this document.

A few comments:
1. Rather than having different rulesets for flagging, I would prefer an INTERMAGNET standard ruleset. This will ease interpretation of the flags for users, instead of having to look up and implement different versions and rulesets.
2. The suggested storage scheme is efficient for continuously flagged data, and I like that additional information is included. However, when multiple single points will be flagged (e.g. spikes), it may be a bit cumbersome. Maybe we can add a second type of flag descriptor where an array of flagged times is provided instead of start and end times.
3. As suggested, flags should remain optional, as they significantly increase workload and should be considered as an "extra" service to the community. The individual institute should decide the level of detail that can be provided.

**Comment by leonro** [comment]
> Rather than having different rulesets for flagging, I would prefer an INTERMAGNET standard ruleset. This will ease interpretation of the flags for users, instead of having to look up and implement different versions and rulesets.

I agree. Nevertheless, rule-sets will probably change with time as most of our work does. Therefore it is not bad to have at least some versioning already included from the beginning. 

> The suggested storage scheme is efficient for continuously flagged data, and I like that additional information is included. However, when multiple single points will be flagged (e.g. spikes), it may be a bit cumbersome. Maybe we can add a second type of flag descriptor where an array of flagged times is provided instead of start and end times.

One could think about am additional layer containing flag specifcations/descriptions as a unique key and a list of individual time steps for which this description apply. That would be easy to be incorporated and very efficient when dealing with many spikes.

**Comment by SimonFlower** [comment]
I would like to see this discussion extended to others responsible for data analysis and quality control at institutes. I think there is another view about data production, which is that the experts in the field (those responsible for the observatory) should produce the best data set they are able to for users without requiring further work by users to determine the data. I don't say that I subscribe to that view, but I think it may be one that comes up and we should extend the discussion to a wider audience in order to have the debate.

Part of what helps me with these types of decision is to consider what we are trying to achieve. Who are the users who would be helped by this additional information and what are their needs, for example? I think answering that question would inform this discussion. Do we already have users for this type of quality control information?

Is an existing data format the right place for this type of information? It will be relatively easy to adapt our CDF format to carry the extra information, but not IAGA-2002. It would make the information easier to distribute with all the formats we use (and require less work updating formats) if the new information was presented in a separate file.

**Comment by leonro** [comment]
Putting all that information into a separate file is simple and can easily be done. 
From a persons perspective responsible for data analysis, quality control and archiving:
What I usually do at our observatory is to archive raw data along with flagging information, all meta information and associated basevalues, functional parameters of adopted baselines in a single CDF structure, for each instrument. Thus I archive raw data and from the contained additional information I can reconstruct basically every analysis step up to finally submitted definitive data. We currently have a project dealing with lightning strikes. Having the appropriate flagging information available makes it very easy to extract specific data from the archive...

**Comment by sputnik-a** [comment]
> Is an existing data format the right place for this type of information? It will be relatively easy to adapt our CDF format to carry the extra information, but not IAGA-2002. It would make the information easier to distribute with all the formats we use (and require less work updating formats) if the new information was presented in a separate file.

Although having a separate file would be an easy thing to do, I would prefer to have the additional information included in a single file along with the data for the simple reason to have them as one inseparable unit. 

In my opinion, we would not have to include the flagging information in IAGA-2002 files at all. Potential users would have to adapt their software tools in any case, even when distributing the flagging information in a separate file. In the best case, this may even foster the usage of the INTERMAGNET cdf-file format ;)

Still, a hybrid solution could be to have a separate flagging file, and having the information included in CDF-files as well - but that's probably no less extra work ....

**Comment by sputnik-a** [comment]
> I would like to see this discussion extended to others responsible for data analysis and quality control at institutes. I think there is another view about data production, which is that the experts in the field (those responsible for the observatory) should produce the best data set they are able to for users without requiring further work by users to determine the data. I don't say that I subscribe to that view, but I think it may be one that comes up and we should extend the discussion to a wider audience in order to have the debate.

Good point. Probably this implies that we don't want to include all data and all flags as used internally at an observatory. For example, @leonro categories one and three should not be delivered at all, such that the provided data "as is", is still the best possible one. In this case, the flagging information would indeed be an extra service that provides additional information and interpretation (@leonro categories 0,2,4). Pity for lightning strikes then ....

**Comment by leonro** [comment]
Well, as said already, it depends on the use-case. The last point is perfectly valid. INTERMAGNET should provide geomagnetic data and thus it is perfectly fine to remove local disturbances such as lightning strikes, cars, etc. As a service you might provide some infos on geomagnetic signatures (e.g. in categories 0,2,4) within the data. If I am performing studies on local disturbances like lightning strikes I will contact the observatory and not check INTERMAGNET. Anyway, the IMAGCDF format is very useful also for storing such data locally in the observatories archive and thus such possibility of keeping flagging info is certainly a benefit.

**Comment by SimonFlower** [comment]
@CharlesBlais @leonro @vmaury-ipgp This issue has not had any activity for some time. Is there a way in which this work can be taken forward, or, if not, should we close this issue (remembering that it can be reopened at a later date)?

**Comment by bgeels-USGS** [comment]
To provide an update from the USGS on this topic: One of our developers (Alexandra Wernle) has recently begun work on a flagging system for use in our GIN. If Intermagnet is still interested in storing or using flags, then we should be able to export our flags in a format that is compatible with what @leonro posted.
